src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,528 @@
+		/// <returns>A dependent context for the session.</returns>
+		protected virtual ITransactionContext CreateDependentContext(ISessionImplementor dependentSession, ITransactionContext mainContext)
+		{
+			return new DependentContext(mainContext);
+		}
+
+		/// <inheritdoc />
+		public override bool IsInActiveSystemTransaction(ISessionImplementor session)
+			=> session.TransactionContext?.IsInActiveTransaction ?? false;
+
+		/// <inheritdoc />
+		public override void ExecuteWorkInIsolation(ISessionImplementor session, IIsolatedWork work, bool transacted)
+		{
+			using (var tx = new TransactionScope(TransactionScopeOption.Suppress))
+			{
+				base.ExecuteWorkInIsolation(session, work, transacted);
+				tx.Complete();
+			}
+		}
+
+		/// <summary>
+		/// Transaction context for enlisting a session with a system <see cref="System.Transactions.Transaction"/>.
+		/// It is meant for being the concrete class enlisted in the transaction.
+		/// </summary>
+		public class SystemTransactionContext : ITransactionContext, IEnlistmentNotification
+		{
+			/// <summary>
+			/// The transaction in which this context is enlisted.
+			/// </summary>
+			protected internal virtual System.Transactions.Transaction AmbientTransaction { get; }
+			/// <inheritdoc />
+			public bool ShouldCloseSessionOnSystemTransactionCompleted { get; set; }
+			/// <inheritdoc />
+			public bool IsInActiveTransaction { get; protected set; } = true;
+			/// <inheritdoc />
+			public virtual bool CanFlushOnSystemTransactionCompleted => _useConnectionOnSystemTransactionEvents;
+
+			private readonly ISessionImplementor _sessionImplementor;
+			private readonly bool _useConnectionOnSystemTransactionEvents;
+			private volatile SemaphoreSlim _semaphore;
+			private volatile bool _locked;
+			private readonly AsyncLocal<bool> _bypassWait = new AsyncLocal<bool>();
+
+			/// <summary>
+			/// Default constructor.
+			/// </summary>
+			/// <param name="sessionImplementor">The session to enlist with the transaction.</param>
+			/// <param name="transaction">The transaction into which the context will be enlisted.</param>
+			/// <param name="useConnectionOnSystemTransactionEvents">See <see cref="Cfg.Environment.UseConnectionOnSystemTransactionEvents"/>.</param>
+			public SystemTransactionContext(
+				ISessionImplementor sessionImplementor,
+				System.Transactions.Transaction transaction,
+				bool useConnectionOnSystemTransactionEvents)
+			{
+				_sessionImplementor = sessionImplementor;
+				var clone = transaction.Clone();
+				AmbientTransaction = clone;
+				clone.TransactionCompleted += TransactionCompleted;
+				_useConnectionOnSystemTransactionEvents = useConnectionOnSystemTransactionEvents;
+			}
+
+			/// <inheritdoc />
+			public virtual void Wait()
+			{
+				if (_isDisposed)
+					return;
+				if (!_locked && AmbientTransaction.TransactionInformation.Status != TransactionStatus.Active)
```

Comment:
When we have not yet locked the context, we are still quite early in the transaction "end of life". I have run a lot of tests since the time I am working on that, and never got a failure here. But why not refactor that access in a catching helper and use it just to stay on the safe side.